SPLASH/src/ImGuiWindows/Chat/ChatWindow.cpp

```diff
@@ -0,0 +1,353 @@
+
+#include "../SPLASH/src/game/events/TextInputEvent.h"
+#include "../SPLASH/src/game/events/NetworkChatEvent.h"
+#include "../SPLASH/src/game/events/NetworkJoinedEvent.h"
+#include "../SPLASH/src/game/events/NetworkDisconnectEvent.h"
+#include "Sail/events/types/NetworkPlayerChangedTeam.h"
+#include "Sail/events/types/WindowResizeEvent.h"
+
+ChatWindow::ChatWindow(bool showWindow) :
+	m_message(""),
+	m_timeSinceLastMessage(0.0f),
+	m_fadeTime(-1.0f),
+	m_fadeThreshold(-1.0f),
+	m_backgroundOpacityMul(0.4f),
+	m_scrollToBottom(false),
+	m_messageLimit(30),
+	m_messageSizeLimit(30),
+	m_retainFocus(true)
+{
+	m_app = Application::getInstance();
+	m_settings = &m_app->getSettings();
+	m_imguiHandler = m_app->getImGuiHandler();
+	m_eventDispatcher = &EventDispatcher::Instance();
+	m_network = &NWrapperSingleton::getInstance();
+
+
+	m_eventDispatcher->subscribe(Event::Type::CHATSENT, this);
+	m_eventDispatcher->subscribe(Event::Type::NETWORK_CHAT, this);
+	m_eventDispatcher->subscribe(Event::Type::NETWORK_JOINED, this);
+	m_eventDispatcher->subscribe(Event::Type::NETWORK_DISCONNECT, this);
+	m_eventDispatcher->subscribe(Event::Type::NETWORK_PLAYER_CHANGED_TEAM, this);
+	m_eventDispatcher->subscribe(Event::Type::WINDOW_RESIZE, this);
+
+
+
+
+
+	m_standaloneButtonflags = ImGuiWindowFlags_NoCollapse |
+		ImGuiWindowFlags_NoResize |
+		ImGuiWindowFlags_NoMove |
+		ImGuiWindowFlags_NoNav |
+		ImGuiWindowFlags_NoBringToFrontOnFocus |
+		ImGuiWindowFlags_NoTitleBar |
+		ImGuiWindowFlags_AlwaysAutoResize |
+		ImGuiWindowFlags_NoSavedSettings |
+		ImGuiWindowFlags_NoBackground;
+	m_backgroundOnlyflags = ImGuiWindowFlags_NoCollapse |
+		ImGuiWindowFlags_NoResize |
+		ImGuiWindowFlags_NoMove |
+		ImGuiWindowFlags_NoNav |
+		ImGuiWindowFlags_NoTitleBar |
+		ImGuiWindowFlags_NoSavedSettings;
+
+
+}
+
+ChatWindow::~ChatWindow() {
+	m_eventDispatcher->unsubscribe(Event::Type::CHATSENT, this);
+	m_eventDispatcher->unsubscribe(Event::Type::NETWORK_CHAT, this);
+	m_eventDispatcher->unsubscribe(Event::Type::NETWORK_JOINED, this);
+	m_eventDispatcher->unsubscribe(Event::Type::NETWORK_DISCONNECT, this);
+	m_eventDispatcher->unsubscribe(Event::Type::NETWORK_PLAYER_CHANGED_TEAM, this);
+	m_eventDispatcher->unsubscribe(Event::Type::WINDOW_RESIZE, this);
+}
+
+bool ChatWindow::onEvent(const Event& event) {
+
+
+	switch (event.type) {
+				
+	case Event::Type::CHATSENT:						onMyTextInput((const ChatSent&)event); break;
+	case Event::Type::NETWORK_CHAT:					onRecievedText((const NetworkChatEvent&)event); break;
+	case Event::Type::NETWORK_JOINED:				onPlayerJoined((const NetworkJoinedEvent&)event); break;
+	case Event::Type::NETWORK_DISCONNECT:			onPlayerDisconnected((const NetworkDisconnectEvent&)event); break;
+	case Event::Type::NETWORK_PLAYER_CHANGED_TEAM:	onPlayerTeamChanged((const NetworkPlayerChangedTeam&)event); break;
+	case Event::Type::WINDOW_RESIZE:				onWindowResize((const WindowResizeEvent&)event); break;
+	
+
+	default:
+		break;
```

Comment:
Could have SAIL_ERROR when the default case is run, not that important however. You may skip this suggestion if you want <3